collect_app/src/main/java/org/odk/collect/android/external/FormUriActivity.kt

```diff
@@ -51,130 +65,41 @@ class FormUriActivity : ComponentActivity() {
-    }
-
-    private fun assertValidUri(): Boolean {
-        val isUriValid = intent.data?.let {
-            val uriMimeType = contentResolver.getType(it)
-            if (uriMimeType == null) {
-                return@let false
-            } else {
-                return@let uriMimeType == FormsContract.CONTENT_ITEM_TYPE || uriMimeType == InstancesContract.CONTENT_ITEM_TYPE
+    private val formUriViewModel by viewModels<FormUriViewModel> {
+        object : ViewModelProvider.Factory {
+            override fun <T : ViewModel> create(modelClass: Class<T>, extras: CreationExtras): T {
+                return FormUriViewModel(
+                    intent.data,
+                    scheduler,
+                    projectsRepository,
+                    projectsDataService,
+                    contentResolver,
+                    formsRepositoryProvider,
+                    instanceRepositoryProvider,
+                    resources
+                ) as T
             }
-        } ?: false
-
-        return if (!isUriValid) {
-            displayErrorDialog(getString(org.odk.collect.strings.R.string.unrecognized_uri))
-            false
-        } else {
-            true
         }
     }
 
-    private fun assertFormExists(): Boolean {
-        val uri = intent.data!!
-        val uriMimeType = contentResolver.getType(uri)
-
-        val doesFormExist = if (uriMimeType == FormsContract.CONTENT_ITEM_TYPE) {
-            formsRepositoryProvider.get().get(ContentUriHelper.getIdFromUri(uri))?.let {
-                File(it.formFilePath).exists()
-            } ?: false
-        } else {
-            instanceRepositoryProvider.get().get(ContentUriHelper.getIdFromUri(uri))?.let {
-                if (!File(it.instanceFilePath).exists()) {
-                    Analytics.log(AnalyticsEvents.OPEN_DELETED_INSTANCE)
-                    InstanceDeleter(instanceRepositoryProvider.get(), formsRepositoryProvider.get()).delete(it.dbId)
-                    displayErrorDialog(getString(org.odk.collect.strings.R.string.instance_deleted_message))
-                    return false
-                }
-
-                val candidateForms = formsRepositoryProvider.get().getAllByFormIdAndVersion(it.formId, it.formVersion)
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        DaggerUtils.getComponent(this).inject(this)
+        setContentView(R.layout.circular_progress_indicator)
 
-                if (candidateForms.isEmpty()) {
-                    val version = if (it.formVersion == null) {
-                        ""
-                    } else {
-                        "\n${getString(org.odk.collect.strings.R.string.version)} ${it.formVersion}"
+        formUriViewModel.error.observe(this) {
+            if (it == null) {
+                if (savedInstanceState != null) {
+                    if (!savedInstanceState.getBoolean(FORM_FILLING_ALREADY_STARTED)) {
+                        startForm()
```

Comment:
Right but because I made things load in the background, I ended up having to build the logic from scratch (moving from sync to async code). There's probably a simpler `if-else` that I haven't spotted. Do you have a suggestion?

Reply:
Nice! The `false` needs to be `true`, but otherwise that works!